sdk/keyvault/internal/challenge_policy.go

```diff
@@ -0,0 +1,189 @@
+//go:build go1.16
+// +build go1.16
+
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+package internal
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+)
+
+const headerAuthorization = "Authorization"
+const bearerHeader = "Bearer "
+
+type KeyVaultChallengePolicy struct {
+	// mainResource is the resource to be retrieved using the tenant specified in the credential
+	mainResource *ExpiringResource
+	cred         azcore.TokenCredential
+	transport    policy.Transporter
+	scope        *string
+	tenantID     *string
+}
+
+func NewKeyVaultChallengePolicy(cred azcore.TokenCredential, t policy.Transporter) *KeyVaultChallengePolicy {
+	return &KeyVaultChallengePolicy{
+		cred:         cred,
+		transport:    t,
+		mainResource: NewExpiringResource(acquire),
+	}
+}
+
+func (k *KeyVaultChallengePolicy) Do(req *policy.Request) (*http.Response, error) {
+	as := acquiringResourceState{
+		p:   k,
+		req: req,
+	}
+
+	if k.scope == nil || k.tenantID == nil {
+		// First request, get both to get the token
+		challengeReq, err := k.getChallengeRequest(req)
+		if err != nil {
+			return nil, err
+		}
+
+		challengeResp, err := k.transport.Do(challengeReq)
+		if err != nil {
+			return nil, err
+		}
+
+		err = k.findScopeAndTenant(challengeResp)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	tk, err := k.mainResource.GetResource(as)
+	if err != nil {
+		return nil, err
+	}
+
+	if token, ok := tk.(*azcore.AccessToken); ok {
+		req.Raw().Header.Set(headerAuthorization, fmt.Sprintf("%s %s", bearerHeader, token.Token))
+	}
+
+	// try the request
+	resp, err := req.Next()
+	if err != nil {
+		return nil, err
+	}
+
+	// If it fails and has a 401, try it with a new token
+	if resp.StatusCode == 401 {
+		// Check for a new auth policy
+		err := k.findScopeAndTenant(resp)
+
+		// Error parsing challenge, doomed to fail. Return
+		if err != nil {
+			return nil, err
+		}
+
+		tk, err := k.mainResource.GetResource(as)
```

Comment:
Could this return the token Key Vault just rejected?

Reply:
It can in which case either the tenant is no longer correct or the scope is invalid. At this point we have no way to authenticate the request and should fail.

Reply:
When we get a 401 we know the cached token is invalid, so we should discard it to prevent the client from using it again.

Reply:
Taken care of, I instantiate a new ExpiringResource in that situation (if that's the wrong way to handle it let me know).

Reply:
Update: Jeff and I fixed this by adding a `Reset` to the `ExpiringResource`